### PR TITLE
Fix no-unused-prop-types setState updater 

### LINKED ISSUE #1506: Invalid no-unused-prop-types catch
Hi,

Try this code:
```
import React from 'react';
import PropTypes from 'prop-types';

class Test extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            foo: false,
        };
    }

    onFoo() {
        this.setState((prevState, props) => {
            props.doSomething();
            return {
                foo: true,
            };
        });
    }

    render() {
        return (
            <div>{this.state.foo}</div>
        );
    }
}

Test.propTypes = {
    doSomething: PropTypes.func.isRequired,
};

export default Test;
```
You will get an error on line `doSomething: PropTypes.func.isRequired,` despite that `doSomething` is used inside the async version of `setState()`. Yep, I know, it's a strange usage, but it should be handled imho ;)

Thanks!

### PR DESCRIPTION
**Pull request by petersendidit**
Handle when props are used in the setState update callback

Fix #1506

### CODE REVIEW

File: tests/lib/rules/no-unused-prop-types.js [hunk 1]
```diff
@@ -2111,6 +2111,24 @@ ruleTester.run('no-unused-prop-types', rule, {
       parser: 'babel-eslint',
       options: [{skipShapeProps: false}]
     }, {
+      // issue #1506
+      code: [
+        'class MyComponent extends React.Component {',
+        '  onFoo() {',
+        '    this.setState((prevState, props) => {',
+        '      props.doSomething();',
```
Comment (ljharb) [inline tests/lib/rules/no-unused-prop-types.js]: does this work independent of what "props" is called?

**Comment by jseminck** [comment]
Does this case work with the prop-types rule? If not: the code is really similar and it should be possible to apply the same logic there.

File: tests/lib/rules/no-unused-prop-types.js (hunk 1 above)
Comment (petersendidit) [reply tests/lib/rules/no-unused-prop-types.js]: Not yet, working a better version that will handle more cases. This code doesn't ] handle destructuring or anything like that.

Comment (ljharb) [reply tests/lib/rules/no-unused-prop-types.js]: I think it's important that it handles that from the get-go.

**Comment by petersendidit** [comment]
Should handle this case much better now. Added a test case for when the second argument to the setState updater is not named `props` and is not destructured. Let me know if there are other test cases that should be added.

**Comment by jseminck** [comment]
What about the prop-types rule? Does it also need this functionality?

**Comment by ljharb** [comment]
Yes, I think it does.

**Pull request merged**